core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PrioritizeUtilizationExecutionSchedule.java

```diff
@@ -0,0 +1,561 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler.policy;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.trino.execution.scheduler.StageExecution;
+import io.trino.execution.scheduler.StageExecution.State;
+import io.trino.sql.planner.PlanFragment;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.ExchangeNode;
+import io.trino.sql.planner.plan.IndexJoinNode;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.PlanFragmentId;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.PlanVisitor;
+import io.trino.sql.planner.plan.RemoteSourceNode;
+import io.trino.sql.planner.plan.SemiJoinNode;
+import io.trino.sql.planner.plan.SpatialJoinNode;
+import org.jgrapht.DirectedGraph;
+import org.jgrapht.EdgeFactory;
+import org.jgrapht.alg.StrongConnectivityInspector;
+import org.jgrapht.graph.DefaultDirectedGraph;
+import oshi.annotation.concurrent.GuardedBy;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.trino.execution.scheduler.StageExecution.State.FLUSHING;
+import static io.trino.execution.scheduler.StageExecution.State.RUNNING;
+import static io.trino.execution.scheduler.StageExecution.State.SCHEDULED;
+import static io.trino.sql.planner.plan.AggregationNode.Step.FINAL;
+import static io.trino.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static io.trino.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+/**
+ * Schedules stages choosing to order to provide the best resource utilization.
```

Comment:
> provide the best resource utilization.

This is misleading. Resource utilization is governed by many other factors unrelated to the order in which stages are scheduled.

What this strategy avoid is scheduling stages that will immediately block waiting for another stage to start producing data. In a way, if I'm understanding the code correctly, it schedules stages that form full non-blocking pipelines in the order in which they are able to make progress.

Reply:
Yup. That's how it works. Do you want a suggestion for a better name? I would use phased execution policy, but it's already taken

Reply:
Hmm.. maybe we could just take the name "phased execution" and call the previous one "legacy phased execution". If this is better in every way, we'll eventually get rid of the old one, anyway.

Incidentally, back when we added phased execution, the original intent was to eventually take advantage of knowledge of which operators were "blocking" vs which ones could make progress in a pipelined manner for the purpose of scheduling stages. We could think of this PR as finally realizing that, but implementing it in a way that provides an escape hatch for the old behavior in case of bugs.

Reply:
renamed in https://github.com/trinodb/trino/pull/10455